ledger/ledgercore/persistedacctdata.go

```diff
@@ -116,3 +117,458 @@ var EncodedMaxAppLocalStates = basics.EncodedMaxAppLocalStates
 
 // EncodedMaxAppParams is a copy from basics package to resolve deps in msgp-generated file
 var EncodedMaxAppParams = basics.EncodedMaxAppParams
+
+// NumAssetHoldings returns number of assets in the account
+func (pad PersistedAccountData) NumAssetHoldings() int {
+	if pad.ExtendedAssetHolding.Count > 0 {
+		return int(pad.ExtendedAssetHolding.Count)
+	}
+	return len(pad.AccountData.Assets)
+}
+
+func (gd *AssetsHoldingGroupData) update(ai int, hodl basics.AssetHolding) {
+	gd.Amounts[ai] = hodl.Amount
+	gd.Frozens[ai] = hodl.Frozen
+}
+
+func (gd *AssetsHoldingGroupData) delete(ai int) {
+	copy(gd.Amounts[ai:], gd.Amounts[ai+1:])
+	gd.Amounts = gd.Amounts[:len(gd.Amounts)-1]
+
+	copy(gd.Frozens[ai:], gd.Frozens[ai+1:])
+	gd.Frozens = gd.Frozens[:len(gd.Frozens)-1]
+
+	if ai == 0 {
+		gd.AssetOffsets = gd.AssetOffsets[1:]
+		gd.AssetOffsets[0] = 0
+	} else if ai == len(gd.AssetOffsets)-1 {
+		gd.AssetOffsets = gd.AssetOffsets[:len(gd.AssetOffsets)-1]
+	} else {
+		gd.AssetOffsets[ai+1] += gd.AssetOffsets[ai]
+		copy(gd.AssetOffsets[ai:], gd.AssetOffsets[ai+1:])
+		gd.AssetOffsets = gd.AssetOffsets[:len(gd.AssetOffsets)-1]
+	}
+}
+
+// GetHolding returns AssetHolding from group data by asset index ai
+func (gd AssetsHoldingGroupData) GetHolding(ai int) basics.AssetHolding {
+	return basics.AssetHolding{Amount: gd.Amounts[ai], Frozen: gd.Frozens[ai]}
+}
+
+// GetHolding returns AssetHolding from group data by asset index ai
+func (g AssetsHoldingGroup) GetHolding(ai int) basics.AssetHolding {
+	return g.groupData.GetHolding(ai)
+}
+
+// Encode returns msgp-encoded group data
+func (g AssetsHoldingGroup) Encode() []byte {
+	return protocol.Encode(&g.groupData)
+}
+
+// GetGroupData returns group data. Used in tests only
+func (g AssetsHoldingGroup) GetGroupData() AssetsHoldingGroupData {
+	return g.groupData
+}
+
+// Update an asset holding by index
+func (g *AssetsHoldingGroup) Update(ai int, holdings basics.AssetHolding) {
+	g.groupData.update(ai, holdings)
+}
+
+// Loaded return a boolean flag indicated if the group loaded or not
+func (g AssetsHoldingGroup) Loaded() bool {
+	return g.loaded
+}
+
+// Load sets a group data value in the group
+func (g *AssetsHoldingGroup) Load(gd AssetsHoldingGroupData) {
+	g.groupData = gd
+	g.loaded = true
+}
+
+func (g *AssetsHoldingGroup) delete(ai int) {
```

Comment:
Add some comment regarding the "meaning" of the `ai` argument. Specifically, that it's *not* an `AssetIndex` but rather an index into this `AssetHoldingGroup`